proposals/interceptors-issues-2024-01.md

```diff
@@ -0,0 +1,147 @@
+
+Ideally if an analyzer needs to know if an interceptor is being used, it could simply make a public API call to ask. This adds an extra decision that component authors need to make. For some use cases, such as rename-refactoring, it doesn't make sense to rename the interceptor method--instead the original method is what needs to be renamed. We think that in most cases, the original method is what should be analyzed, but it should still be noted as a pit that people may fall into.
+
+That public API could look something like the following:
+
+```diff
+ namespace Microsoft.CodeAnalysis;
+
+ public static class CSharpExtensions
+ {
++    public static IMethodSymbol? GetInterceptorMethod(this SemanticModel model, InvocationExpressionSyntax invocation, CancellationToken cancellationToken = default);
+ }
+```
+
+Note that if we do something like this, there will be a deep need to discover interceptors, and incrementally update compilations containing interceptors, with as little binding work as possible. It will be non-tenable perf-wise if, for example, we had to bind all the attributes in a compilation in order to decide what the interceptor for a given invocation is. We may need to store some additional information in the declaration tree, which can be incrementally updated, in order to address this.
+
+# Future (possibly post-C# 13)
+
+## Brittleness of line/column approaches
+
+While we want interceptor info to be readily available to components in the IDE, we also want the IDE to be able to provide the best possible function while not running source generators on every "keystroke plus delay". Even after our work on incremental source generators, they are still quite expensive to run, and we will need a strategy such as "regenerate on save" or "regenerate on build" in order to claw back IDE performance.
+
+The desire for less-frequent source generation is at odds with the desire for interceptor info in IDE components, as interceptors are currently designed.
+
+For example, we want the ILLinker analyzer's results to be as useful as possible, and ideally to not include *spurious transient warnings* while the user is making edits. For example, if a user inserts some white spaces in a method body which contains intercepted calls, we don't want the ILLinker analyzer to report that some non-intercepted call is using an unsupported method. But this requires source generators to run. This *strongly* pushes the ILLinker analyzer to run *only as frequently as interceptor source generators do*.
+
+This would be an undesirable degradation in the user experience. Instead of being able to give users confidence that their mistakes "Native AOT-wise" will be caught very quickly by the tooling and shown to them as they type, they instead are looking at longer delays, or even not being told something is wrong until they save or hit build.
```

Comment:
This section feels a bit off to me. This problem is not specific to interceptors or the ILLinker analyzers. It is a more general tension between when we run generators (any generator) and analyzers. If generators are run less frequently then analyzers that depend on their outputs will suffer. 

Think this may be better as a note of concern on the IDE discussion on strategies to delay running of generators

https://github.com/dotnet/roslyn/issues/67441